src/htm/regions/DatabaseOutRegion.cpp

```diff
@@ -0,0 +1,355 @@
+/* ---------------------------------------------------------------------
+ * HTM Community Edition of NuPIC
+ * Copyright (C) 2013, Numenta, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ * --------------------------------------------------------------------- */
+
+/** @file
+ * Implementation for DatabaseOutRegion class
+ */
+
+#include <iostream>
+#include <list>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+#include <stdio.h>
+
+#include <htm/engine/Output.hpp>
+#include <htm/engine/Input.hpp>
+#include <htm/engine/Region.hpp>
+#include <htm/engine/Spec.hpp>
+#include <htm/regions/DatabaseOutRegion.hpp>
+#include <htm/utils/Log.hpp>
+
+#define MAX_NUMBER_OF_INPUTS 10 // maximal number of inputs/scalar streams in the database
```

Comment:
we try to avoid defines, if possible. could you use `const UInt MAX_NUM... = 10; `

Reply:
...as my prev. comment about "how users can modify this", it is already fine like this 
- if users really need more than 10 inputs, they can change it here and recompile the code. 
- would it be possible to make the MAX_NUM_OF_INPUTS part of the region's Spec, ie making it easier to change?

Reply:
Hmmm, I wonder if you need it at all.
These are used to declare the inputs so you can use links with them.  I have not thought this through but what if you used one input and used the Fan-In capability of the Link.  The links for data for all fields would be connected to the same input.  The Fan-In facility will append the buffers together and are passed into the region as one big array. 
Don't know....probably would be hard to separate them out into fields inside the region.   Probably not a good idea.
The point is there might be another way to solve this problem.

Reply:
> I see this parameter IS in the Spec. It is the number spec fields to create. The Spec is static, used before the Region is created 

so the other way, can we "read" it from the Spec and get it to the field here?

Reply:
> Hmmm, I wonder if you need it at all.
These are used to declare the inputs so you can use links with them. 

you need it for the fields in the DB (?)

Reply:
> you need it for the fields in the DB (?)

We need to declare inputs so we can get data for each field of the record being written to the database.  Ideally the names of those inputs should be the names of the db fields being stored.   But since those names can be anything and could be different for each instance of the region, they could not be in the Spec.  The Spec must be static and the same for all instances of the region, set at compile time.   So that is why he has used generic names in the Spec.

I was trying to think of a way where the names of the inputs which correspond to DB fields, and their type, could be set during region creation if not even at runtime like it is in SQL.

Reply:
> I was trying to think of a way where the names of the inputs which correspond to DB fields, and their type, could be set during region creation if not even at runtime like it is in SQL.

Hmm yes i am thinking the same way, but unless we define new link datatype like tuple, we can't dynamically add the inputs.. I want to keep this simple. I have no idea now how to implement this better.